pkitesting/src/main/java/io/netty/pkitesting/X509Bundle.java

```diff
@@ -0,0 +1,349 @@
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.pkitesting;
+
+import io.netty.util.internal.EmptyArrays;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.security.KeyPair;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.CertificateEncodingException;
+import java.security.cert.CertificateException;
+import java.security.cert.TrustAnchor;
+import java.security.cert.X509Certificate;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.List;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.TrustManagerFactory;
+
+/**
+ * A certificate bundle is a private key and a full certificate path, all the way to the root certificate.
+ * The bundle offers ways of accessing these, and converting them into various representations.
+ */
+public final class X509Bundle {
+    private final X509Certificate[] certPath;
+    private final X509Certificate root;
+    private final KeyPair keyPair;
+
+    /**
+     * Construct a bundle from a given certificate path, root certificate, and {@link KeyPair}.
+     * @param certPath The certificate path, starting with the leaf certificate.The path can end either with the
+     * root certificate, or the intermediate certificate signed by the root certificate.
+     * @param root The self-signed root certificate.
+     * @param keyPair The key pair.
+     */
+    public X509Bundle(X509Certificate[] certPath, X509Certificate root, KeyPair keyPair) {
+        if (certPath.length > 1 && certPath[certPath.length - 1].equals(root)) {
+            this.certPath = Arrays.copyOf(certPath, certPath.length - 1);
+        } else {
+            this.certPath = certPath.clone();
+        }
+        this.root = root;
+        this.keyPair = keyPair;
+    }
+
+    /**
+     * Construct a bundle for a certificate authority.
+     * @param root The self-signed root certificate.
+     * @param keyPair The key pair.
+     * @return The new bundle.
+     */
+    public static X509Bundle fromRootCertificateAuthority(X509Certificate root, KeyPair keyPair) {
+        X509Bundle bundle = new X509Bundle(new X509Certificate[]{root}, root, keyPair);
```

Comment:
Should we add null checks for root and keyPair ? Same comment also goes for others methods